### PR TITLE
feat: auth context api proposal

### PR DESCRIPTION
**Pull request by erickzanardo**
## Status

**IN DEVELOPMENT**

## Description

This PR adds a proposal for a new API that will allow users of the auth middleware to have access to the `RequestContext`, which will allow them to make use of the dependency injection.

The proposed change is implemented with the idea of not introducing a breaking change, it will allow the user to chose between the previous API and the new one. Here is an example of the usage of the new api:

```dart
        bearerAuthentication<User>(
          userFromContext: (token, context) async {
            final sessionRepository = context.read<SessionRepository>();
            final session = await sessionRepository.sessionFromToken(token);

            if (session != null) {
              final userRepository = context.read<UserRepository>();
              return userRepository.findUserById(session.userId);
            }

            return null;
          },
        ),
```
As you can note, the new API is super similar to the previous one, it only adds a new argument (the RequestContext) in the received function.

At first, I thought that we could choose to deprecate the previous api, but while I was working on this it could make sense to have both options since in some approaches of authentication, like JWT, there might not exist the need to access the request context, since the token is self contained and the decryption keys would probably be inject in the application via an environment variable. So seems convenient to offer the user an option where they don't need to declare an argument without needing it.

The only thing that I am not happy about this change is the name `userFromContext`, I feel it don't really describe what it really does. But at the same time `userFromTokenAndContext`/`userFromCredentialsAndContext` seems a mouthful (I miss methods overload :( ). So I am open to suggestions.

I have not changed any docs or tests yet as I seek for feedback before doing so. Once we are all in agreement with the change, I will do so.

## Type of Change

- [ ] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: packages/dart_frog_auth/lib/src/dart_frog_auth.dart [hunk 1]
```diff
@@ -47,27 +47,42 @@ Future<bool> _defaultApplies(RequestContext context) async => true;
 /// [RequestContext]. If the function returns false, the middleware will not
 /// apply to the route and the call will have authentication validation.
 Middleware basicAuthentication<T extends Object>({
-  required Future<T?> Function(String, String) userFromCredentials,
+  Future<T?> Function(String, String)? userFromCredentials,
```
Comment (erickzanardo) [inline packages/dart_frog_auth/lib/src/dart_frog_auth.dart:50]: Note: Although I changed this to not be required anymore, this will not be a breaking change since anyone who was using the middleware will already have this informed, so the assert bellow will not throw.

**Comment by renancaraujo** [comment]
This is the less compromising approach. As an addition, I would add `@deprecated` to the previous arguments

File: packages/dart_frog_auth/lib/src/dart_frog_auth.dart [hunk 2]
```diff
@@ -90,21 +105,37 @@ Middleware basicAuthentication<T extends Object>({
 /// [RequestContext]. If the function returns false, the middleware will not
 /// apply to the route and the call will have no authentication validation.
 Middleware bearerAuthentication<T extends Object>({
-  required Future<T?> Function(String) userFromToken,
+  Future<T?> Function(String)? userFromToken,
+  Future<T?> Function(String, RequestContext)? userFromContext,
```
Comment (alestiago) [inline packages/dart_frog_auth/lib/src/dart_frog_auth.dart:109]: **Nit:**

I would provide the names of the arguments so when implemented they are kept consistent. Same applies to others.
```suggestion
  Future<T?> Function(String token, RequestContext context)? userFromContext,
```

File: packages/dart_frog_auth/lib/src/dart_frog_auth.dart [hunk 3]
```diff
@@ -47,27 +47,42 @@ Future<bool> _defaultApplies(RequestContext context) async => true;
 /// [RequestContext]. If the function returns false, the middleware will not
 /// apply to the route and the call will have authentication validation.
 Middleware basicAuthentication<T extends Object>({
-  required Future<T?> Function(String, String) userFromCredentials,
+  Future<T?> Function(String, String)? userFromCredentials,
+  Future<T?> Function(String, String, RequestContext)? userFromContext,
```
Comment (alestiago) [inline packages/dart_frog_auth/lib/src/dart_frog_auth.dart:51]: I agree that we need to work on the naming. I feel that after reading `userFromCredentials`, I get the feeling that `userFromContext` does not provide any credentials, instead these credentials are in the context.

In addition, to keep consistency with Flutter (for example, ListView.builder's itemBuilder) I would always make the context the first parameter.

We could, once again, take inspiration from Flutter and rename the `userFromContext` to `userBuilder`.  The builder receives the `context` and the credentials as parameters. Personally, I think in the long term we can deprecate the old API.

File: docs/docs/advanced/authentication.md [hunk 4]
```diff
@@ -122,14 +125,18 @@ Handler middleware(Handler handler) {
       .use(requestLogger())
       .use(
         bearerTokenAuthentication<User>(
-          userFromToken: userRepository.fetchFromAccessToken,
-        ),
+          readUser: (context, token) {
+            final userRepository = context.read<UserRepository>();
+            return userRepository.fetchFromAccessToken(token);
+          }
+        ).build(),
```
Comment (renancaraujo) [inline docs/docs/advanced/authentication.md:132]: I would split the docs to a different pr to be merged after the package release

File: packages/dart_frog_auth/lib/src/dart_frog_auth.dart [hunk 5]
```diff
@@ -47,27 +47,53 @@ Future<bool> _defaultApplies(RequestContext context) async => true;
 /// [RequestContext]. If the function returns false, the middleware will not
 /// apply to the route and the call will have authentication validation.
 Middleware basicAuthentication<T extends Object>({
-  required Future<T?> Function(String, String) userFromCredentials,
+  @Deprecated(
+    'Deprecated in favor of readUser. '
+    'This might be removed in future versions',
```
Comment (renancaraujo) [inline packages/dart_frog_auth/lib/src/dart_frog_auth.dart:52]:
```suggestion
    'This will be removed in future versions',
```

File: packages/dart_frog_auth/lib/src/dart_frog_auth.dart [hunk 6]
```diff
@@ -90,21 +116,41 @@ Middleware basicAuthentication<T extends Object>({
 /// [RequestContext]. If the function returns false, the middleware will not
 /// apply to the route and the call will have no authentication validation.
 Middleware bearerAuthentication<T extends Object>({
-  required Future<T?> Function(String) userFromToken,
+  @Deprecated(
+    'Deprecated in favor of readUser. '
+    'This might be removed in future versions',
```
Comment (renancaraujo) [inline packages/dart_frog_auth/lib/src/dart_frog_auth.dart:121]:
```suggestion
    'This will be removed in future versions',
```

**Review by renancaraujo — APPROVED** [review]

File: packages/dart_frog_auth/lib/src/dart_frog_auth.dart (hunk 3 above)
Comment (alestiago) [reply packages/dart_frog_auth/lib/src/dart_frog_auth.dart]: I understand what about: `userReader`?

File: docs/docs/advanced/authentication.md (hunk 4 above)
Comment (erickzanardo) [reply docs/docs/advanced/authentication.md]: Docs reverted and added a PR only for it!

**Review by renancaraujo — APPROVED** [review]

**Review by renancaraujo — APPROVED** [review]
